src/sema/eval.rs

```diff
@@ -196,3 +201,186 @@ pub fn eval_const_rational(
+                    *loc,
+                    ty.clone(),
+                    *unchecked,
+                    Box::new(left),
+                    Box::new(right),
+                )
+            }
+        }
+
+        Expression::Power(loc, ty, unchecked, left, right) => {
+            let left = eval_constants_in_expression(left, ns, results);
+            let right = eval_constants_in_expression(right, ns, results);
+
+            if let (Expression::NumberLiteral(_, _, left), Expression::NumberLiteral(_, _, right)) =
+                (&left, &right)
+            {
+                results.push(left.pow(right.to_u32().unwrap()));
+                Expression::NumberLiteral(*loc, ty.clone(), left.pow(right.to_u32().unwrap()))
+            } else {
+                Expression::Power(
+                    *loc,
+                    ty.clone(),
+                    *unchecked,
+                    Box::new(left),
+                    Box::new(right),
+                )
+            }
+        }
+
+        Expression::ShiftLeft(loc, ty, left, right) => {
+            let left = eval_constants_in_expression(left, ns, results);
+            let right = eval_constants_in_expression(right, ns, results);
+
+            if let (Expression::NumberLiteral(_, _, left), Expression::NumberLiteral(_, _, right)) =
+                (&left, &right)
+            {
+                results.push(left.shl(right.to_u32().unwrap()));
+                Expression::NumberLiteral(*loc, ty.clone(), left.shl(right.to_u32().unwrap()))
+            } else {
+                Expression::ShiftLeft(*loc, ty.clone(), Box::new(left), Box::new(right))
+            }
+        }
+
+        Expression::ShiftRight(loc, ty, left, right, _) => {
+            let left = eval_constants_in_expression(left, ns, results);
+            let right = eval_constants_in_expression(right, ns, results);
+
+            if let (Expression::NumberLiteral(_, _, left), Expression::NumberLiteral(_, _, right)) =
+                (&left, &right)
+            {
+                results.push(left.shr(right.to_u32().unwrap()));
+                Expression::NumberLiteral(*loc, ty.clone(), left.shr(right.to_u32().unwrap()))
+            } else {
+                Expression::ShiftLeft(*loc, ty.clone(), Box::new(left), Box::new(right))
+            }
+        }
+        Expression::NumberLiteral(_, _, n) => {
+            results.push(n.clone());
+            expr.clone()
+        }
+
+        Expression::Builtin(.., args) => {
+            for args_iter in args {
+                verify_result(args_iter, ns, &args_iter.loc());
+            }
+
+            expr.clone()
+        }
+
+        Expression::InternalFunctionCall { args, .. }
+        | Expression::ExternalFunctionCall { args, .. } => {
+            for args_iter in args.clone() {
+                verify_result(&args_iter, ns, &args_iter.loc());
+            }
+
+            expr.clone()
+        }
+        _ => expr.clone(),
```

Comment:
There are many expressions missing from this list, for example `new D(arg1, arg2)`, and calls with named arguments `foo({a: 1, b: 1})`, method calls etc. I think it might be better to use `fn recurse()` on `Expression`, see https://github.com/hyperledger/solang/blob/main/src/sema/ast.rs#L802